### PR TITLE
Feature/fid

### LINKED ISSUE #11: Unify interfaces for FID and KID
**Is your feature request related to a problem? Please describe.**
Now these two measure have different high-level interfaces and work only with passed extracted features. Make it better.

**Describe the solution you'd like**
Create interfaces that allows to specify feature extractor, accept different types of collections of images etc.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by zakajd**
Partially closes #11 

To be discussed:
For now after computing dataset features they are converted into numpy arrays, to be used in `scipy.linalg.sqrtm` function. There is exist implementation of the same function on PyTorch (see: https://github.com/msubhransu/matrix-sqrt or https://github.com/pytorch/pytorch/issues/25481 ).
So maybe it's worth it to rewrite `scipy.linalg.sqrtm` in the future to get some increase in speed?

### CODE REVIEW

File: photosynthesis_metrics/fid.py [hunk 1]
```diff
@@ -56,15 +51,99 @@ def __compute_fid(mu1: float, sigma1: np.ndarray, mu2: float, sigma2: np.ndarray
...
+        """
+
+        assert predicted_features.shape[1] == target_features.shape[1], \
+            f"Features dimensionalities should match, otherwise it won't be possible to correctly compute statistics. \
+                Got {predicted_features.shape[1]} and {target_features.shape[1]}"
+        assert predicted_features.dims == 2, f"Predicted features must have shape  (N_samples, encoder_dim), got {predicted_features.dims}"
+        assert target_features.dims == 2, f"Target features must have shape  (N_samples, encoder_dim), got {target_features.dims}"
+
+        score = compute_fid(predicted_features.numpy(), target_features.numpy())
+        return torch.Tensor(score)
```
Comment (snk4tr) [inline photosynthesis_metrics/fid.py]: Please consider a conventional way to convert numpy arrays to torch tensors by using `torch.from_numpy()`. The difference is that `torch.Tensor()` always creates a float32 tensor, but `torch.from_numpy()` preserves the initial data type of the numpy array that was passed. 

For example:

```python
score = np.array([[1, 2, 3], [4, 5, 6]])
print(torch.from_numpy(score).dtype)
print(torch.Tensor(score).dtype)

>>> torch.int64
>>> torch.float32
```

Or

```python
score = np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float64)
print(torch.from_numpy(score).dtype)
print(torch.Tensor(score).dtype)

>>> torch.float64
>>> torch.float32
```

File: photosynthesis_metrics/fid.py [hunk 2]
```diff
@@ -56,15 +51,99 @@ def __compute_fid(mu1: float, sigma1: np.ndarray, mu2: float, sigma2: np.ndarray
...
+
+    def _compute_feats(self, loader, feature_extractor : Callable = None, device: str = 'cuda') -> torch.Tensor:
+        r"""Generate low-dimensional image desciptors to be used for computing MSID score. 
+        Args:
+            loader: Batch of images generator. Batche required to be 4D, channels first (N,C,H,W).
+            feature_extractor: model used to generate image features, if None use `InceptionNet V2` model
+            out_features: size of `feature_extractor` output
+            device: Device on which to compute inference of the model
+        """
+
```
Comment (snk4tr) [inline photosynthesis_metrics/fid.py]: Formatting: no blank line here

File: photosynthesis_metrics/fid.py [hunk 3]
```diff
@@ -56,15 +51,99 @@ def __compute_fid(mu1: float, sigma1: np.ndarray, mu2: float, sigma2: np.ndarray
...
+        assert predicted_features.shape[1] == target_features.shape[1], \
+            f"Features dimensionalities should match, otherwise it won't be possible to correctly compute statistics. \
+                Got {predicted_features.shape[1]} and {target_features.shape[1]}"
+        assert predicted_features.dims == 2, f"Predicted features must have shape  (N_samples, encoder_dim), got {predicted_features.dims}"
+        assert target_features.dims == 2, f"Target features must have shape  (N_samples, encoder_dim), got {target_features.dims}"
+
+        score = compute_fid(predicted_features.numpy(), target_features.numpy())
+        return torch.Tensor(score)
+
```
Comment (snk4tr) [inline photosynthesis_metrics/fid.py]: Formatting: no blank line here

File: photosynthesis_metrics/fid.py [hunk 4]
```diff
@@ -56,15 +51,99 @@ def __compute_fid(mu1: float, sigma1: np.ndarray, mu2: float, sigma2: np.ndarray
...
+            loader: Batch of images generator. Batche required to be 4D, channels first (N,C,H,W).
+            feature_extractor: model used to generate image features, if None use `InceptionNet V2` model
+            out_features: size of `feature_extractor` output
+            device: Device on which to compute inference of the model
+        """
+
+        if feature_extractor is None:
+            print('WARNING: default feature extractor (InceptionNet V2) is used.')
+            feature_extractor = InceptionV3()
+            # self.out_features = ... # Inception V2() out features
```
Comment (snk4tr) [inline photosynthesis_metrics/fid.py]: What is this line for? Is it a self-reminder or TODO?

File: photosynthesis_metrics/fid.py [hunk 5]
```diff
@@ -56,15 +51,99 @@ def __compute_fid(mu1: float, sigma1: np.ndarray, mu2: float, sigma2: np.ndarray
...
+        else:
+            assert callable(feature_extractor), f"Feature_extractor must be callabel. Got {type(feature_extractor)}"
```
Comment (snk4tr) [inline photosynthesis_metrics/fid.py]: Shouldn't it be not just callable, but instance of `torch.nn.Module`?

File: photosynthesis_metrics/fid.py [hunk 6]
```diff
@@ -56,15 +51,99 @@ def __compute_fid(mu1: float, sigma1: np.ndarray, mu2: float, sigma2: np.ndarray
...
+
+        feature_extractor.eval()
+        
+        if out_features is None:
```
Comment (snk4tr) [inline photosynthesis_metrics/fid.py]: `out_features` is not defined above. Is it ok?

File: photosynthesis_metrics/fid.py [hunk 7]
```diff
@@ -56,15 +51,99 @@ def __compute_fid(mu1: float, sigma1: np.ndarray, mu2: float, sigma2: np.ndarray
...
+
+        total_feats = []
+        for batch in loader:
+            N = batch.shape[0]
+            batch = batch.float().to(device)
+
+            # Getting features
+            batch_feats = feature_extractor(batch).view(N, )
+            total_feats.append(batch_feats)
+        return torch.Tensor(total_feats)
```
Comment (snk4tr) [inline photosynthesis_metrics/fid.py]: See the comment about `torch.Tensor` above.

**Comment by snk4tr** [comment]
@zakajd, it seems that it would be valuable to introduce a common interface for metrics that use feature extractors (FID, KID, MSID etc.) as we discussed before.

**Comment by zakajd** [comment]
One important thing in code for `_compute_feats` method.
Now InceptionV3 model by default returns list with one tensor in it - features from the last layer of the model before fully-connected part. There is also an option to return several features from different parts of the model. Same functionality can be found in other repos (from [segmentation_models.pytorch](https://github.com/qubvel/segmentation_models.pytorch/blob/master/segmentation_models_pytorch/encoders/inceptionv4.py) or  [pytorch_toolbelt](https://github.com/BloodAxe/pytorch-toolbelt/blob/develop/pytorch_toolbelt/zoo/segmentation.py)). 

But here lies a problem, because all metrics expect to get features of fixed size and  there I can't think of a general way to flatten multiple features into one vector, so that it works in all cases. 
So, I added an assert 
```
features = feature_extractor(images)
assert len(features) == 1, \
    f"feature_encoder must return list with features from one layer. Got {len(features)}"
```

This can be discussed.

**Comment by zakajd** [comment]
@snk4tr  Check this out. Looks like it's almost ready to be merged.

File: photosynthesis_metrics/fid.py [hunk 8]
```diff
@@ -52,19 +46,62 @@ def __compute_fid(mu1: float, sigma1: np.ndarray, mu2: float, sigma2: np.ndarray
...
+
+class FID(BaseFeatureMetric):
+    r"""Creates a criterion that measures Frechet Inception Distance score for two datasets of images
+    See https://arxiv.org/abs/1706.08500 for reference.
+    """
+    def __init__(self):
+        super(FID, self).__init__()
+        self.compute = compute_fid
+
+    def forward(self, predicted_features: torch.Tensor, target_features: torch.Tensor) -> torch.Tensor:
```
Comment (snk4tr) [inline photosynthesis_metrics/fid.py]: Wrong type annotation. This forward is annotated such that it returns `torch.Tensor`, but `compute_fid` from which you obtain the actual score returns `float`. That is why this method always returns `float`.

```suggestion
    def forward(self, predicted_features: torch.Tensor, target_features: torch.Tensor) -> float:
```

File: photosynthesis_metrics/fid.py [hunk 9]
```diff
@@ -1,45 +1,39 @@
...
-
-    predicted_features = feature_extractor(predicted_images)
-    target_features = feature_extractor(target_images)
-
-    # TODO: `compute_fid` works with np.arrays, but need to work with torch.Tensors. Refactor that
-    return compute_fid(predicted_features, target_features)
-
+from photosynthesis_metrics.utils import BaseFeatureMetric, _validate_features
 
 def __compute_fid(mu1: float, sigma1: np.ndarray, mu2: float, sigma2: np.ndarray, eps=1e-6) -> float:
```
Comment (snk4tr) [inline photosynthesis_metrics/fid.py]: Please change type annotations for `mu1` and `mu2` here. Now they are set to be float, but you alway pass results of `compute_statistics` here, which are `Tuple[np.ndarray, np.ndarray]`

```suggestion
def __compute_fid(mu1: np.ndarray, sigma1: np.ndarray, mu2: np.ndarray, sigma2: np.ndarray, eps=1e-6) -> float:
```

File: photosynthesis_metrics/fid.py [hunk 10]
```diff
@@ -1,45 +1,39 @@
...
+    The Frechet Inception Distance between two multivariate Gaussians X_predicted ~ N(mu_1, sigm_1)
+    and X_target ~ N(mu_2, sigm_2) is
+        d^2 = ||mu_1 - mu_2||^2 + Tr(sigm_1 + sigm_2 - 2*sqrt(sigm_1*sigm_2)).
+    Args:
+        mu1: mean of activations calculated on predicted samples
+        sigma1: covariance matrix over activations calculated on predicted samples
+        mu2: mean of activations calculated on target samples
+        sigma2: covariance matrix over activations calculated on target samples
+        eps: offset constant. used if sigma_1 @ sigma_2 matrix is singular
+    """
```
Comment (snk4tr) [inline photosynthesis_metrics/fid.py]: Update docs: what is returned?

File: photosynthesis_metrics/kid.py [hunk 11]
```diff
@@ -1,69 +1,84 @@
 import sys
```
Comment (snk4tr) [inline photosynthesis_metrics/kid.py]: Delete unused import

File: photosynthesis_metrics/fid.py [hunk 12]
```diff
@@ -1,45 +1,39 @@
...
+PyTorch implementation of Frechet Inception Distance (FID score)
+Reference:
+    Martin Heusel et al. "GANs Trained by a Two Time-Scale Update Rule Converge to a Local Nash Equilibrium",
+    https://arxiv.org/abs/1706.08500
+Credits:
+    https://github.com/hukkelas/pytorch-frechet-inception-distance/
+    https://github.com/mseitzer/pytorch-fid
+"""
+
+from typing import Callable, Tuple
```
Comment (snk4tr) [inline photosynthesis_metrics/fid.py]: Delete unused import `Callable`

File: photosynthesis_metrics/fid.py [hunk 13]
```diff
@@ -1,45 +1,39 @@
...
 
 import numpy as np
-
+import torch
 from scipy import linalg
-from typing import Tuple, Optional
 from torch import nn
```
Comment (snk4tr) [inline photosynthesis_metrics/fid.py]: Delete unused import

File: photosynthesis_metrics/fid.py [hunk 14]
```diff
@@ -1,45 +1,39 @@
...
-        print('WARNING: default feature extractor (InceptionNet V2) is used.')
-        feature_extractor = InceptionV3()
-
-    predicted_features = feature_extractor(predicted_images)
-    target_features = feature_extractor(target_images)
-
-    # TODO: `compute_fid` works with np.arrays, but need to work with torch.Tensors. Refactor that
-    return compute_fid(predicted_features, target_features)
-
+from photosynthesis_metrics.utils import BaseFeatureMetric, _validate_features
```
Comment (snk4tr) [inline photosynthesis_metrics/fid.py]: Delete unused import `_validate_features`

File: photosynthesis_metrics/kid.py [hunk 15]
```diff
@@ -1,69 +1,84 @@
...
+from photosynthesis_metrics.utils import BaseFeatureMetric
+
+def compute_polynomial_mmd_averages(
+    x : np.ndarray,
+    y : np.ndarray,
+    n_subsets : int = 50,
+    subset_size : int = 1000,
+    ret_var : bool = True,
+    **kernel_args
+    ) -> Union[List[np.ndarray], Tuple[List[np.ndarray], List[np.ndarray]]]:
```
Comment (snk4tr) [inline photosynthesis_metrics/kid.py]: Wrong return type annotation now
```suggestion
    ) -> Union[Tuple[ndarray, ndarray], ndarray]:
```

File: photosynthesis_metrics/kid.py [hunk 16]
```diff
@@ -1,69 +1,84 @@
...
+        y: Samples from data distribution. Shape (N_samples, data_dim), dtype: torch.float32 in range 0 - 1
+        degree: Degree of a polynomial functions used in kernels. Default: 3
+        gamma: Kernel parameter. See paper for details
+        coef0: Kernel parameter. See paper for details
+        var_at_m: Kernel variance. Default is `None`
+        ret_var: whether to return variance after the distance is computed.
                        This function will return Tuple[float, float] in this case.
-
     Returns:
         float or Tuple[float, float]: KID score and variance (optional).
```
Comment (snk4tr) [inline photosynthesis_metrics/kid.py]: Return type in the docstring does not correspond to the one in the type annotation. Better just remove it
```suggestion
        KID score and variance (optional).
```

File: photosynthesis_metrics/kid.py [hunk 17]
```diff
@@ -1,69 +1,84 @@
...
-def _mmd2_and_variance(K_XX, K_XY, K_YY, unit_diagonal=False,
-                       mmd_est='unbiased', block_size=1024,
-                       var_at_m=None, ret_var=True):
+def _mmd2_and_variance(
+    K_XX : np.ndarray,
+    K_XY : np.ndarray,
+    K_YY : np.ndarray,
+    unit_diagonal : bool = False,
+    mmd_est : str = 'unbiased',
+    block_size : int = 1024,
```
Comment (snk4tr) [inline photosynthesis_metrics/kid.py]: Remove unused argument

File: photosynthesis_metrics/kid.py [hunk 18]
```diff
@@ -124,29 +139,78 @@ def _mmd2_and_variance(K_XX, K_XY, K_YY, unit_diagonal=False,
...
+    Reference:
+        Demystifying MMD GANs https://arxiv.org/abs/1801.01401 
+    """
+    def __init__(
+        self,
+        degree : int = 3,
+        gamma : Optional[float] = None,
+        coef0 : int = 1,
+        var_at_m : Optional[int] = None,
+        ret_var : bool = True
```
Comment (snk4tr) [inline photosynthesis_metrics/kid.py]: Remove unused arguments (do not forget to change docstring after removal)

File: photosynthesis_metrics/utils.py [hunk 19]
```diff
@@ -0,0 +1,73 @@
+from typing import Callable, Optional, Union, Tuple, List
```
Comment (snk4tr) [inline photosynthesis_metrics/utils.py]: Remove unused imports (all of them)

File: photosynthesis_metrics/utils.py [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+
+import torch
+import numpy as np
```
Comment (snk4tr) [inline photosynthesis_metrics/utils.py]: Remove unused import

File: photosynthesis_metrics/utils.py [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+
+        total_feats = []
+        for batch in loader:
+            images = batch['images']
+            N = images.shape[0]
+            images = images.float().to(device)
+
+            # Get features
+            features = feature_extractor(images)
+            assert len(features) == 1, \
```
Comment (snk4tr) [inline photosynthesis_metrics/utils.py]: As we discussed before, I am fine with this change, but please add a description of this behaviour somewhere in docstring if possible

File: requirements.txt [hunk 22]
```diff
@@ -1,3 +1,4 @@
 torch==1.4.0
 scipy==1.4.1
+sklearn
```
Comment (snk4tr) [inline requirements.txt]: Please specify your version. Thus we can be sure that the lib will work after the latest version changes

File: tests.ipynb [hunk 23]
```diff
@@ -0,0 +1,472 @@
+{
```
Comment (snk4tr) [inline tests.ipynb]: No notebooks in the lib. Move test to scripts that will be runnable with pytest or remove the notebook

File: tests/test_kid.py [hunk 24]
```diff
@@ -0,0 +1,54 @@
+import pytest
+import torch
+
+from photosynthesis_metrics import KID, compute_polynomial_mmd
+from photosynthesis_metrics.feature_extractors.fid_inception import InceptionV3
```
Comment (snk4tr) [inline tests/test_kid.py]: Remove unused import

**Pull request merged**